### PR TITLE
Added support for Azure in git_submodules

### PR DESCRIPTION
**Pull request by mburumaxwell**
Support `git_submodules` from Azure DevOps. All the code was present but not set up.

Changes in #6304 and #6306 helped a great deal too in handling submodule URLs for Azure DevOps:

```bash
[submodule "another-repo"]
	path = another-repo
	url = https://contoso@dev.azure.com/contoso/MyProject/_git/another-repo
```

The suffix `.git` does not in Azure DevOps. For GitHub and GitLab, it works with and without the suffix hence removed it just when getting the `git-upload-pack` in `service_pack_uri(...)`. For `dev.azure.com` and `*.visualstudio.com`, this suffix is skipped.

Fixes: #6340 
Fixes: #5756

### CODE REVIEW

**Comment by mburumaxwell** [comment]
@jeffwidman I realize this will not pass CI because of the removal of the `.git` suffix. It seems to affect quite much.
This affects the URLs generated in #6321 . Fetching in #6343 would also be affected (though it gracefully failsover).

How you suggest going about that?

**Comment by mburumaxwell** [comment]
@jeffwidman 
Seeing as to how complex the removal of the `.git` suffix is, I have opted to exclude it only for `dev.azure.com` and `*.visualstudio.com` hostnames as there is no other way to tell if we should. This will not work for Azure DevOps Server but maybe that is a fight for another day.

**Comment by jeffwidman** [comment]
👋thanks for digging into this. I'll probably be able to take a look this coming week. One of my focuses the next few weeks is catching up on our backlog of PRs.

**Review by jeffwidman — COMMENTED** [review]
So the code looks reasonable given that today Azure doesn't respond to `*.git` URI's, unlike all other source control providers.

However, before merging this, I'd like to reach out to the ADO team and get their perspective... they might be willing to add support for these URI's given that all the other platforms already have this support so they're essentially lagging behind.

I realize that even if that gets on their roadmap, it won't solve your problem today, so we'll still probably want to merge this, but I'd feel better if we first have a conversation with them... they may also know some setting that we're unaware of to handle this.

Given that Microsoft owns GitHub, it shouldn't be too difficult for me to track down who I need to speak with there... I've pinged a few people so will keep you posted.

File: common/lib/dependabot/git_metadata_fetcher.rb
Comment (jeffwidman) [inline common/lib/dependabot/git_metadata_fetcher.rb:173]:
```suggestion
      # TODO: Unlike the other providers (GitHub, GitLab, BitBucket), as of 2023-01-18 Azure DevOps does not support the ".git" suffix. It will return a 404.
      # So skip adding ".git" if looks like an ADO URI. 
      # There's no access to the source object here, so have to check the URI instead.
      # Even if we had the current source object, the URI may be for a dependency hosted elsewhere.
      # Unfortunately as a consequence, urls pointing to Azure DevOps Server will not work.
      # Only alternative is to remove the addition of ".git" suffix since the other providers
      # (GitHub, GitLab, BitBucket) work with or without the suffix.
      # That change has other ramifications, so it'd be better if Azure started supporting ".git" like all the other providers.
```

**Comment by jeffwidman** [comment]
@lachie83 do you happen to know who on the ADO team I can chat with about ☝️? Happy to chat privately, my GitHub email/slack handle is same as my GitHub username.

**Comment by mburumaxwell** [comment]
> So the code looks reasonable given that today Azure doesn't respond to `*.git` URI's, unlike all the other popular source control providers.
> 
> Before merging this, I'd like to reach out to the ADO team and get their perspective... they might be willing to add support to get to feature parity. Or they might know some setting/URL that we're unaware of to handle this.
> 
> I realize that even if that gets on their roadmap, it won't solve your problem today, so after an initial conversation to ensure we're not overlooking something obvious then we'll still probably move forward with merging this and then open a tracking ticket in our backlog to remind us to check down the road if they added support.
> 
> Given that Microsoft owns GitHub, it shouldn't be too difficult for me to track down who I need to speak with there... I've pinged a few people so will keep you posted.

I am also not too thrilled about the `.git` suffix not being supported on Azure DevOps while everyone else seems to do it. I would be really glad if they actually did add support for it but I will keep my expectations in check. It should ideally be easier for you to tell. According to https://docs.antora.org/antora/latest/playbook/git-suffix/#ensure-git-suffix-key, other people are also affected.

OTOH, [GitHub](https://github.com/dependabot/dependabot-core), [Gitlab](https://gitlab.com/dependabot-gitlab/dependabot), and [Bitbucket](https://bitbucket.org/atlassian_tutorial/helloworld) support URLs without the `.git` suffix. I haven't gotten to study the implications of not having the suffix as regards the internal workings of this repo but could not having the suffix for all actually be an option/solution?

**Comment by deivid-rodriguez** [comment]
If `.git`-less URLs work everywhere, we may do that as well. It's a higher risk change, sure, but better in terms of future maintenance. Thoughts @jeffwidman?

**Comment by deivid-rodriguez** [comment]
> But I do care about the code/UX experience being intuitive... and to me these URI's are more predictable/expected with .git suffixes... But I'm only n=1, so others may have a different perspective.

I would hope we can make the code good enough so that its readability does not depend on the specific shape of this URI. And its complexity should be about the same as now (less than the approach in this PR). Not sure how it affects the UX either, you mean in logs and so on?

Anyways, let's definitely hear from ADO people first!

**Comment by mburumaxwell** [comment]
This also fixes #5756

**Comment by waltervos** [comment]
I'm not sure if it's appreciated, but I just wanted to let you know that this change would mean a lot to my team :)

**Review by jeffwidman — APPROVED** [review]
Sorry for the delay, got buried here.

I just emailed the ADO team, but will probably take a few weeks to sync up with them.

In the meantime, my vote is we should ship this as-is for now... my thinking is that it's a small self-contained change, easy to back out later if ADO ends up adding support for `.git` suffix down the road.

Until we talk to them though, I hesitate to spend further time exploring dropping the `.git` suffix. @mburumaxwell already looked into it a bit and says it touches a number of things. And right now the only thing we're forgoing is Azure DevOps Server support. Plus, I actually _like_ the signaling of the `.git` suffix... it feels intuitive to me. 😄 

So that's why I vote we punt for now on that and figure out next steps after that ADO conversation happens.

What say you @deivid-rodriguez ?

**Comment by deivid-rodriguez** [comment]
Ok!

**Comment by mburumaxwell** [comment]
I'm in total agreement too.

**Pull request merged**
